runtime/src/main/java/org/corfudb/runtime/view/stream/StreamAddressSpace.java

```diff
@@ -0,0 +1,288 @@
+            return;
+        }
+
+        throw new RuntimeException("Couldn't seek " + streamId + " to address " + address);
+    }
+
+    @Override
+    public long getMax() {
+        if (maxInd == -1) {
+            return Address.NON_ADDRESS;
+        } else {
+            return addresses.get(maxInd);
+        }
+    }
+
+    @Override
+    public long getMin() {
+        if (addresses.size() == 0) {
+            return Address.NON_ADDRESS;
+        } else {
+            return addresses.get(0);
+        }
+    }
+
+    @Override
+    public long getLastAddressSynced() {
+        return getMax();
+    }
+
+    @Override
+    @Synchronized
+    public long next() {
+        if (currInd + 1 > maxInd) {
+            return Address.NON_ADDRESS;
+        } else {
+            currInd++;
+            return addresses.get(currInd);
+        }
+    }
+
+    @Override
+    @Synchronized
+    public long previous() {
+        if (currInd - 1 >= 0) {
+            currInd--;
+            return addresses.get(currInd);
+        } else {
+            currInd = -1;
+            return Address.NON_ADDRESS;
+        }    }
+
+    @Override
+    public List<Long> remainingUpTo(long limit) {
+        List<Long> res = new ArrayList<>(100);
+        while (hasNext() && addresses.get(currInd + 1) <= limit) {
+            res.add(addresses.get(currInd + 1));
+            currInd++;
+        }
+        return res;
+    }
+
+    @Override
+    public void addAddresses(List<Long> addresses) {
+        this.addresses.addAll(addresses);
+        Collections.sort(this.addresses);
+        maxInd += addresses.size();
+    }
+
+    @Override
+    public long getCurrentPointer() {
+        if (currInd == -1) {
+            return Address.NON_ADDRESS;
+        } else {
+            return addresses.get(currInd);
+        }
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (currInd < maxInd) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean hasPrevious() {
+        if (currInd > 0 && currInd < addresses.size()) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public void removeAddresses(long upperBound) {
+        int removedAddresses = 0;
+        int index = this.addresses.indexOf(upperBound);
+        if (index != -1) {
+            for (int i = index; i >= 0; i--){
+                removedAddresses ++;
+                this.addresses.remove(i);
+            }
+
+            if (currInd <= index) {
+                // Reset current index to first entry in the stream
+                currInd = -1;
+            } else {
+                currInd = currInd - removedAddresses;
+            }
+
+            maxInd = this.addresses.size() - 1;
+        }
+    }
+
+    @Override
+    public void syncUpTo(long globalAddress, long newTail, long lowerBound,
+                         Function<Long, ILogData> readFn) {
```

Comment:
why do you need readFn ? AddressSpaceView.read already defines the reading protocol.